File: showcase/go/showcase_integration_test.go
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package showcase_integration

import (
	"context"
	"flag"
	"io"
	"log"
	"os"
	"strings"
	"testing"
	"time"

	showcase "cloud.google.com/go/showcase/apiv1alpha2"
	durationpb "github.com/golang/protobuf/ptypes/duration"
	genprotopb "github.com/googleapis/gapic-showcase/server/genproto"
	"google.golang.org/api/iterator"
	"google.golang.org/api/option"
	spb "google.golang.org/genproto/googleapis/rpc/status"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

var client *showcase.EchoClient

func TestMain(m *testing.M) {
	flag.Parse()

	conn, err := grpc.Dial("localhost:7469", grpc.WithInsecure())
	if err != nil {
		log.Fatal(err)
	}
	clientOpt := option.WithGRPCConn(conn)
	client, err = showcase.NewEchoClient(context.Background(), clientOpt)
	if err != nil {
		log.Fatal(err)
	}

	os.Exit(m.Run())
}

func TestEcho(t *testing.T) {
	t.Parallel()
	content := "hello world!"
	req := &genprotopb.EchoRequest{
		Response: &genprotopb.EchoRequest_Content{
			Content: content,
		},
	}
	resp, err := client.Echo(context.Background(), req)

	if err != nil {
		t.Fatal(err)
	}
	if resp.GetContent() != req.GetContent() {
		t.Errorf("Echo() = %q, want %q", resp.GetContent(), content)
	}
}

func TestEcho_error(t *testing.T) {
	t.Parallel()
	val := codes.Canceled
	req := &genprotopb.EchoRequest{
		Response: &genprotopb.EchoRequest_Error{
			Error: &spb.Status{Code: int32(val)},
		},
	}
	resp, err := client.Echo(context.Background(), req)

	if resp != nil {
		t.Errorf("Echo() = %v, wanted error %d", resp, val)
	}
	status, _ := status.FromError(err)
	if status.Code() != val {
		t.Errorf("Echo() errors with %d, want %d", status.Code(), val)
	}
}

func TestExpand(t *testing.T) {
	t.Parallel()
	content := "The rain in Spain stays mainly on the plain!"
	req := &genprotopb.ExpandRequest{Content: content}
	s, err := client.Expand(context.Background(), req)
	if err != nil {
		t.Fatal(err)
	}
	resps := []string{}
	for {
		resp, err := s.Recv()
		if err == io.EOF {
			break
		}
		if err != nil {
			t.Fatal(err)
		}
		resps = append(resps, resp.GetContent())
	}
	got := strings.Join(resps, " ")
	if content != got {
		t.Errorf("Expand() = %q, want %q", got, content)
	}
}

func TestCollect(t *testing.T) {
	t.Parallel()
	content := "The rain in Spain stays mainly on the plain!"
	s, err := client.Collect(context.Background())
	if err != nil {
		t.Fatal(err)
	}

	for _, str := range strings.Split(content, " ") {
		s.Send(&genprotopb.EchoRequest{
			Response: &genprotopb.EchoRequest_Content{Content: str}})
	}

	resp, err := s.CloseAndRecv()
	if err != nil {
		t.Fatal(err)
	}
	if content != resp.GetContent() {
		t.Errorf("Collect() = %q, want %q", resp.GetContent(), content)
	}
}

func TestChat(t *testing.T) {
	t.Parallel()
	content := "The rain in Spain stays mainly on the plain!"
	s, err := client.Chat(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	for _, str := range strings.Split(content, " ") {
		s.Send(&genprotopb.EchoRequest{
			Response: &genprotopb.EchoRequest_Content{Content: str}})
	}
	s.CloseSend()
	resps := []string{}
	for {
		resp, err := s.Recv()
		if err == io.EOF {
			break
		}
		if err != nil {
			t.Fatal(err)
		}
		resps = append(resps, resp.GetContent())
	}
	got := strings.Join(resps, " ")
	if content != got {
		t.Errorf("Chat() = %q, want %q", got, content)
	}
}

func TestWait(t *testing.T) {
	t.Parallel()
	content := "hello world!"
	req := &genprotopb.WaitRequest{
		ResponseDelay: &durationpb.Duration{Nanos: 100},
		Response: &genprotopb.WaitRequest_Success{
			Success: &genprotopb.WaitResponse{Content: content},
		},
	}
	resp, err := client.Wait(context.Background(), req)
	if err != nil {
		t.Fatal(err)
	}
	if resp.GetContent() != content {
		t.Errorf("Wait() = %q, want %q", resp.GetContent(), content)
	}
}

func TestWait_timeout(t *testing.T) {
	t.Parallel()
	content := "hello world!"
	req := &genprotopb.WaitRequest{
		ResponseDelay: &durationpb.Duration{Seconds: 1},
		Response: &genprotopb.WaitRequest_Success{
			Success: &genprotopb.WaitResponse{Content: content},
		},
	}
	ctx, _ := context.WithTimeout(context.Background(), 20*time.Millisecond)
	resp, err := client.Wait(ctx, req)

	if err == nil {
		t.Errorf("Wait() = %v, want error", resp)
	}
}


func TestPagination(t *testing.T) {
	t.Parallel()
	req := &genprotopb.PaginationRequest{PageSize: 5, MaxResponse: 20}
	iter := client.Pagination(context.Background(), req)

	expected := int32(0)
	for {
		i, err := iter.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			t.Fatal(err)
		}
		if i != expected {
			t.Errorf("Chat() = %d, want %d", i, expected)
		}
		expected++
	}
	if expected != 20 {
		t.Errorf("Pagination expected to see 20 vals")
	}
}

File: showcase/python/tests/showcase_integration_test.py
# Copyright 2018 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Showcase based integration tests."""

import grpc
import pytest

from google import showcase_v1alpha2
from google.api_core import exceptions
from google.rpc import status_pb2
from google.showcase_v1alpha2.proto import echo_pb2
from google.showcase_v1alpha2.gapic.transports import echo_grpc_transport


class TestEchoClient(object):
    channel = grpc.insecure_channel('localhost:7469')
    transport = echo_grpc_transport.EchoGrpcTransport(channel=channel)
    client = showcase_v1alpha2.EchoClient(transport=transport)

    def test_echo(self):
        content = 'hello world'
        response = self.client.echo(content=content)
        assert content == response.content

    def test_echo_exception(self):
        with pytest.raises(exceptions.InternalServerError):
            self.client.echo(
                error=status_pb2.Status(code=13, message='error!!!'))

    def test_expand(self):
        content = 'The rain in Spain stays mainly on the Plain!'
        response = []
        for r in self.client.expand(content=content):
            response.append(r.content)

        assert content == ' '.join(response)

    def test_collect(self):
        content = 'The rain in Spain stays mainly on the Plain!'
        requests = content.split(' ')
        requests = map(lambda s: echo_pb2.EchoRequest(content=s), requests)
        response = self.client.collect(iter(requests))

        assert content == response.content

    def test_chat(self):
        content = 'The rain in Spain stays mainly on the Plain!'
        requests = content.split(' ')
        requests = map(lambda s: echo_pb2.EchoRequest(content=s), requests)
        responses = self.client.chat(iter(requests))
        responses = map(lambda r: r.content, responses)

        assert content == ' '.join(responses)

    def test_wait(self):
        content = 'hello world'
        response = self.client.wait(
            response_delay={'nanos': 500 }, success={'content': content})
        assert content == response.content

    def test_pagination(self):
        expected = 0
        for element in self.client.pagination(20, page_size=5):
            assert element == expected
            expected = expected + 1
